### ISSUE TITLE
[webview_flutter] does not load relative file URLs on a real iOS device 

### ISSUE THREAD

**Issue by ghowen** [flutter/flutter#67747]
## The issue
Due to https://github.com/flutter/flutter/issues/67628 I tried a different approach to viewing local HTML files in a Flutter app. Instead of serving the files through a local webserver (which crashes on iOS), I now add the files to the asset bundle and on app start I copy them to the local document directory on the device. From there I can load the HTML files in a webview using the `file://` schema, with the benefit that images, css etc. that are referenced in the HTML are loaded as well.

This works fine in Android (simulator and real device) as well as in an iOS simulator. But on an iOS device the webview fails to load css files with a relative URL, if the referenced file is in a different directory which is not below the current file.

## Steps to Reproduce

1. Run `flutter create test_webview`.
2. Update the files as follows:  

- Create the following directory structure in the project
````
assets
assets/www
assets/www/text
assets/www/css

````
- Create  `index.html` in `assets/www/text`:

````html
<!DOCTYPE html>
<html >
  <head>
    <title>Test </title>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no"/>
    <link rel="stylesheet" type="text/css" href="../css/test.css"/>
  </head>
  <body>
    <h1>This text is red.</h1>
  </body>
</html>

````
- Create `test.css` in `assets/www/css`:
````
h1 {
    color: red;
}

````

- Update `pubspec.yaml` like this:
````yaml
name: test_webview
description: A new Flutter project.

publish_to: 'none' # Remove this line if you wish to publish to pub.dev
version: 1.0.0+1

environment:
  sdk: ">=2.7.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  cupertino_icons: ^1.0.0
  path: ^1.7.0
  path_provider: ^0.5.0+1
  webview_flutter: ^1.0.0

dev_dependencies:
  flutter_test:
    sdk: flutter

flutter:

  uses-material-design: true

  assets:
     - assets/www/
     - assets/www/text/
     - assets/www/css/

````

- Update `main.dart` like this:
````dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'package:path/path.dart';
import 'package:path_provider/path_provider.dart';
import 'package:webview_flutter/webview_flutter.dart';

Directory documentsDirectory;

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  documentsDirectory = await getApplicationDocumentsDirectory();
  await _copyAssetsToDocuments();
  runApp(MyApp());
}

Future<void> _copyAssetsToDocuments() async {
  final assetsDir = Directory(join(documentsDirectory.path, 'assets'));

  if (assetsDir.existsSync()) {
    assetsDir.deleteSync(recursive: true);
  }

  final manifestContent = await rootBundle.loadString('AssetManifest.json');
  var manifestMap = json.decode(manifestContent);
  var webAssets = manifestMap.keys
      .where((key) => key is String && key.startsWith('assets/www/'));

  for (var asset in webAssets) {
    await _copyFromAssets(asset);
  }
}

Future<void> _writeToFile(ByteData data, String path) {
  final buffer = data.buffer;
  return new File(path)
      .writeAsBytes(buffer.asUint8List(data.offsetInBytes, data.lengthInBytes));
}

Future<void> _copyFromAssets(String pathToAsset) async {
  try {
    var bytes = await rootBundle.load(pathToAsset);
    final destinationPath = join(documentsDirectory.path, pathToAsset);
    final destinationDir = Directory(dirname(destinationPath));

    if (await destinationDir.exists()) {
      await _writeToFile(bytes, destinationPath);
      print(destinationPath);
    } else {
      await destinationDir.create(recursive: true);
      await _writeToFile(bytes, destinationPath);
      print(destinationPath);
    }
  } catch (e) {
    print(e.toString());
  }
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: WebView(
          initialUrl: 'file://' +
              documentsDirectory.path +
              '/assets/www/text/index.html',
          javascriptMode: JavascriptMode.unrestricted,
          onWebResourceError: (error) {
            print(error.description);
          },
        ),
      ),
    );
  }
}


````

- Add this to your `info.plist` in XCode:

````plist
<dict>
	<key>NSAllowsArbitraryLoads</key>
	<true/>
	<key>NSAllowsLocalNetworking</key>
	<true/>
</dict>

````

3. ... Run the app on an iOS simulator and a real iOS device.

**Expected results:** 

The text "This text is red" should be red.

**Actual results:** 

In the iOS simulator the text is red. On a real iOS device the text is black.

<details>
  <summary>Logs</summary>

```
```

```
georg@MM20 test_webview % flutter analyze
Analyzing test_webview...
No issues found! (ran in 1.7s)
```

```
georg@MM20 test_webview % flutter doctor -v
[✓] Flutter (Channel stable, 1.22.1, on Mac OS X 10.15.7 19H2, locale de-PT)
    • Flutter version 1.22.1 at /Users/georg/bin/flutter
    • Framework revision f30b7f4db9 (24 hours ago), 2020-10-08 10:06:30 -0700
    • Engine revision 75bef9f6c8
    • Dart version 2.10.1


[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/georg/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 50.0.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.0

[✓] Connected device (3 available)
    • Nokia 5 3 (mobile)             • N0AA003628K71000589                      • android-arm64 • Android 10 (API 29)
    • sdk gphone x86 (mobile)        • emulator-5554                            • android-x86   • Android 11 (API 30) (emulator)
    • Georgs iPhone 6s Plus (mobile) • 677d8be9c78d54dbce7fc7f97a46c0f81f31322c • ios           • iOS 14.0.1
    ! Error: Georgs iPhone 6s Plus is busy: Copying cache files from device. Xcode will continue when Georgs iPhone 6s Plus is
      finished. (code -10)

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
@ghowen Thanks for the detailed report.
I am able to replicate the issue on iPhone 6s with iOS 14. Although the same code works properly on Samsung A2Core Galaxy device and on iPhone SE simulator.

![67747](https://user-images.githubusercontent.com/67046386/95721248-f0847000-0c8f-11eb-9bf7-3e601d79269a.png)
<img width="371" alt="Screenshot 2020-10-12 at 1 35 19 PM" src="https://user-images.githubusercontent.com/67046386/95721254-f2e6ca00-0c8f-11eb-8b71-8a331959ead4.png">

![IMG-0308](https://user-images.githubusercontent.com/67046386/95721275-faa66e80-0c8f-11eb-83a2-77efabd8eb46.PNG)

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 1.22.1, on Mac OS X 10.15.4 19E2269, locale en-IN)
    • Flutter version 1.22.1 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision f30b7f4db9 (4 days ago), 2020-10-08 10:06:30 -0700
    • Engine revision 75bef9f6c8
    • Dart version 2.10.1

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (2 available)
    • Darshan's iphone (mobile)           •
      21150b119064aecc249dfcfe05e259197461ce23 • ios • iOS 14.0
    • iPhone SE (2nd generation) (mobile) • 6C85835D-FBFD-4AB3-8DE8-B4FAD35E5367
      • ios • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)

• No issues found!

```

</details>

**Comment by ghowen** [comment]
@darshankawar Exactly. It works for me on Android simulator and device as well as iOS simulator, but fails on a real iOS device. This is strange as the iOS simulator should behave the same in that case (as it is not really a hardware related issue). 

My guess is that is has something to with how the plugin on iOS load the `file://` URL. iOS has a "special" method for these kind of URLs called `loadFileURL(_:allowingReadAccessTo:)` where the second parameter defines the base url for all accessible directories. The docs for iOS state that 

> If readAccessURL references a single file, only that file may be loaded by WebKit. If readAccessURL references a directory, files inside that file may be loaded by WebKit.

As far as I can understand the source the plugin does not use this method at all but uses the `load(_ request: URLRequest)` method, which also seems to work with `file://` URL, but not properly.

**Comment by ghowen** [comment]
This has been killing me for the past days, but I just found out that the example above fails to run on Android (simulator and device) if the `compileSdkVersion` and `targetSdkVersion` is `30`. It only works up to version `29`. The error shown when trying to load the local html file is `net::ERR_ACCESS_DENIED`, although all files are still being copied correctly from the assets to the document directory.

**Comment by darshankawar** [comment]
@ghowen 
For api level 30, try by adding this in manifest file: `<uses-permission android:name="android.permission.QUERY_ALL_PACKAGES"/>` try again and see if it works.

**Comment by ghowen** [comment]
@darshankawar Thanks for the idea, but unfortunately it doesn't work. I get the same ERR_ACCESS_DENIED. 

I also tried `<uses-permission android:name="android.permission.INTERNET"/>` and `android:usesCleartextTraffic="true"`, but so far no luck. It's ok for my app to be on `29` for now but if `30` becomes mandatory for the Play store this will be a major issue.

**Comment by mockturtl** [comment]
> the example above fails to run on Android (simulator and device) if the compileSdkVersion and targetSdkVersion is 30. It only works up to version 29.

@ghowen Hi, I saw your Medium article.  

Android 30 has a breaking change :sweat:, described here: https://pub.dev/documentation/url_launcher/latest/url_launcher/canLaunch.html

Related: https://github.com/flutter/flutter/issues/63727#issuecomment-736120631

I haven't tested the `file` scheme myself, but I expect a variation of this will fix it.

```diff
--- a/android/app/src/main/AndroidManifest.xml
+++ b/android/app/src/main/AndroidManifest.xml
@@ -6,7 +6,13 @@
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.my.app">

    <uses-permission android:name="android.permission.INTERNET" />
    ...
-
+    <queries>
+        <intent>
+            <action android:name="android.intent.action.VIEW" />
+            <category android:name="android.intent.category.BROWSABLE" />
+            <data android:scheme="https" />
+        </intent>
+    </queries>
    <!-- io.flutter.app.FlutterApplication is an android.app.Application that
         calls FlutterMain.startInitialization(this); in its onCreate method.
         In most cases you can leave this as-is, but you if you want to provide
         additional functionality it is fine to subclass or reimplement
         FlutterApplication and put your custom class here. -->
    <application
       ...
```

**Comment by ghowen** [comment]
@mockturtl Thanks for the pointer! I will try this when I raise the target SDK to 30 on Android. This reminds me again that I should not mix two issues but open a separate one for each ;) But again, thanks!

**Comment by SerkanColak** [comment]
> > the example above fails to run on Android (simulator and device) if the compileSdkVersion and targetSdkVersion is 30. It only works up to version 29.
> 
> @ghowen Hi, I saw your Medium article.
> 
> Android 30 has a breaking change 😓, described here: https://pub.dev/documentation/url_launcher/latest/url_launcher/canLaunch.html
> 
> Related: [#63727 (comment)](https://github.com/flutter/flutter/issues/63727#issuecomment-736120631)
> 
> I haven't tested the `file` scheme myself, but I expect a variation of this will fix it.
> 
> ```diff
> --- a/android/app/src/main/AndroidManifest.xml
> +++ b/android/app/src/main/AndroidManifest.xml
> @@ -6,7 +6,13 @@
> <manifest xmlns:android="http://schemas.android.com/apk/res/android"
>     package="com.my.app">
> 
>     <uses-permission android:name="android.permission.INTERNET" />
>     ...
> -
> +    <queries>
> +        <intent>
> +            <action android:name="android.intent.action.VIEW" />
> +            <category android:name="android.intent.category.BROWSABLE" />
> +            <data android:scheme="https" />
> +        </intent>
> +    </queries>
>     
>     <application
>        ...
> ```

I've tested it, Works like a charm, 100% guaranteed 👍 Thanks for your solution @mockturtl

**Comment by danagbemava-nc** [comment]
Issue is reproducible  on `master 2.5.0-7.0.pre.65` and `stable 2.2.3` with ``webview_flutter: ^2.0.10` using the code sample provided by author.

On the simulator the text is red as shown below: 

![Simulator Screen Shot - iPhone 12 - 2021-08-13 at 07 32 54](https://user-images.githubusercontent.com/88313112/129321800-4e517514-1d13-4776-8c8f-da1ad3a5f455.png)

but the text is black on the real device. Tested on `IPhone XS` running `iOS 14.6`.
![IMG_2815](https://user-images.githubusercontent.com/88313112/129321894-14ab8f79-f25b-4968-910d-1d71ee23694f.PNG)


<details>
<summary> flutter doctor -v</summary>

```console
[✓] Flutter (Channel master, 2.5.0-7.0.pre.65, on macOS 11.5.1 20G80 darwin-arm,
    locale en-GH)
    • Flutter version 2.5.0-7.0.pre.65 at /Users/nexus/dev/sdks/flutters
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4bd8b288bb (5 hours ago), 2021-08-12 18:02:02 -0700
    • Engine revision 670a681c1d
    • Dart version 2.15.0 (build 2.15.0-6.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.59.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (3 available)
    • Nexus’ IPhone (mobile) • 00008020-001875E83A38002E            • ios
      • iOS 14.6 18F72
    • iPhone 12 (mobile)     • F2592D43-5FE3-46BB-8A2B-A38BA468DCE9 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • Chrome (web)           • chrome                               •
      web-javascript • Google Chrome 92.0.4515.131

• No issues found!
```


```console
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.5.1 20G80 darwin-arm, locale
    en-GH)
    • Flutter version 2.2.3 at /Users/nexus/dev/sdks/flutter
    • Framework revision f4abaa0735 (6 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.59.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (3 available)
    • Nexus’ IPhone (mobile) • 00008020-001875E83A38002E            • ios
      • iOS 14.6
    • iPhone 12 (mobile)     • F2592D43-5FE3-46BB-8A2B-A38BA468DCE9 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • Chrome (web)           • chrome                               •
      web-javascript • Google Chrome 92.0.4515.131

• No issues found!
```
</details>

**Comment by nikhilkWinjit** [comment]
@darshankawar 

**Actual Problem :**
While working on Flutter Application to load Web View. I have separate files for HTML and JavaScript i.e. I'm loading JavaScript file externally from HTML file. But when I try to load the webview on iOS platform, I'm getting the following error.

![Could not create Sandbox extension](https://user-images.githubusercontent.com/89373737/130619950-2b1fe61a-a616-4906-bd86-e32971114bec.png)

**Platforms:**
Mobile Device : iPhone 6s
iOS : 13.0
Flutter : 2.2.3
Plugin : webview_flutter 2.0.10

**Expected Behaviour :**
WebView should be load on iOS platform without any issue.

**Comment by eseidelGoogle** [comment]
My guess is this has to do with sandboxing being different for WebView on an iOS device vs. simulator.  I suspect WebView has callbacks we could hook into to log and see what path it's trying to load.

I'd probably try comparing things discussed here https://stackoverflow.com/questions/6420925/load-resources-from-relative-path-using-local-html-in-uiwebview or in Apple's docs, to how the WebView plugin implements.

For example:
WebView(
          initialUrl: 'file://' +
              documentsDirectory.path +
              '/assets/www/text/index.html',
          javascriptMode: JavascriptMode.unrestricted,
          onWebResourceError: (error) {

It's not clear to me (without looking at the plugin code) where the "root directory" being served is from that call.  If it's just the parent of index.html, I could understand that the OS might not let you escape that root with ../../.

**Comment by eseidelGoogle** [comment]
This is likely the line causing you grief:
https://github.com/flutter/plugins/blob/master/packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FlutterWebView.m#L245

The quick fix is to move the assets you need deeper in the path hierarchy, rather than as parent directories to your initial index.html.

Alternatively, Flutter's WebView would need to give you ability to control what the "root" url iOS restricts loads to.

**Comment by eseidelGoogle** [comment]
FYI @mvanbeusekom who was last to touch this callback and may have thoughts.

**Comment by mvanbeusekom** [comment]
Version 3.0.0 of the webview_flutter plugin supports loading assets (including files that are references using relative paths) using the separate `WebViewController.loadFlutterAssets(String key)` method. 

I have modified the above test case to make use of the new `WebViewController.loadFlutterAssets()` method which works correctly on both the simulator and real device (iPhone 12 mini in my case). Another advantage is that the code necessary becomes a lot simpler:

- Updated `pubspec.yaml` file:
```yaml
name: webview
description: A new Flutter project.
publish_to: 'none'
version: 1.0.0+1

environment:
  sdk: ">=2.15.1 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  webview_flutter: ^3.0.0

dev_dependencies:
  flutter_test:
    sdk: flutter
  flutter_lints: ^1.0.0

flutter:
  uses-material-design: true

  assets:
     - assets/www/
     - assets/www/text/
     - assets/www/css/
```
- Updated `main.dart` file:
```dart
import 'dart:io';
import 'package:flutter/material.dart';

import 'package:webview_flutter/webview_flutter.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: const WebViewExample(),
    );
  }
}

class WebViewExample extends StatefulWidget {
  const WebViewExample({Key? key}) : super(key: key);

  @override
  WebViewExampleState createState() => WebViewExampleState();
}

class WebViewExampleState extends State<WebViewExample> {
  @override
  void initState() {
    super.initState();
    // Enable virtual display.
    if (Platform.isAndroid) WebView.platform = AndroidWebView();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Flutter Demo Home Page'),
      ),
      body: Center(
        child: WebView(
          javascriptMode: JavascriptMode.unrestricted,
          onWebResourceError: (error) {
            print(error.description);
          },
          onWebViewCreated: (controller) =>
              controller.loadFlutterAsset('assets/www/text/index.html'),
        ),
      ),
    );
  }
}
```

The `assets/www/css/test.css` and `assets/www/text/index.html` files are copied from the original post as is. 

One disadvantage is that the current setup will cause a small flicker as it will first load `about:blank` as a `initialUrl` and only load the assets after the `onWebViewCreated` callback is called. We are currently working on refactoring the webview plugin to resolve this issue and allow initialisation of the `WebView` widget separate from the widget tree (see also flutter/flutter#88347).

**Comment by AntonShipunov** [comment]
I have the same problem as the author of the issue.
"I now add the files to the asset bundle and on app start I copy them to the local document directory on the device".
If you also load html files from device, you can try using this method. 

### For webview_flutter: ^3.0.0

1.  You need to use the new method from the WebViewController "loadFile(absoluteFilePath)" instead of "loadUrl(url)"

This method:
`_controller.loadUrl(Uri(scheme: 'file', path: htmlPath).toString());`

Replace to:
`_controller.loadFile(htmlPath);`

2.  Then in the "FlutterWebView.m" file, which is located in the "webview_flutter_wkwebview-2.7.1\ios\Classes\FlutterWebView.m" directory, the following code in line 245 must be replaced:
```
NSURL* baseUrl = [url URLByDeletingLastPathComponent];
[_webView loadFileURL:url allowingReadAccessToURL:baseUrl];
```
baseUrl - this is the directory where the downloaded file is located. Relative paths will work for files that are in this directory.

To extend access to other files, replace to:
```
  NSURL* documentsURL = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] objectAtIndex:0];
  [_webView loadFileURL:url allowingReadAccessToURL:documentsURL];
```

It is work for me.
I hope this helped someone.

**Comment by mvanbeusekom** [comment]
@AntonShipunov why don't you use the `WebViewController.loadFlutterAsset()` method as documented in my previous comment? This method has been specifically implemented to load content directly from the Flutter Assets bundle which means you don't have to copy files to a temporary location. It also means it is not necessary to change the Objective-C implementation.

**Comment by AntonShipunov** [comment]
@mvanbeusekom I have been looking for this solution for several days, and the whole implementation in my app is based on the fact that files are copied to the device.
I wrote this solution if anyone needs to download files from the device, for whatever reason. 

It's great that now it's possible to open local files from Assets that contain relative paths.
I will definitely do code refactoring in the future, thank you.

**Comment by istvanszoboszlai** [comment]
> @AntonShipunov why don't you use the `WebViewController.loadFlutterAsset()` method as documented in my previous comment? This method has been specifically implemented to load content directly from the Flutter Assets bundle which means you don't have to copy files to a temporary location. It also means it is not necessary to change the Objective-C implementation.

In my case relative references from an html loaded with the `WebViewController.loadFlutterAsset()` do not work. I figured out that on Android, I need to reference the files in my html prefixing with `android_asset/flutter_assets/`. I could not figure out what works on iOS. But referencing files without any prefix in the same asset directory does not work on either platform.
Do you have a working example?

**Comment by istvanszoboszlai** [comment]
@mvanbeusekom, do you have a working example on usage of `loadFlutterAsset()`? For me only the copying to temp dir approach worked.

**Comment by AntonShipunov** [comment]
@istvanszoboszlai 
Can you show your WebView implementation example, folder structure inside assets, and pubspec.yaml file?

**Comment by mvanbeusekom** [comment]
Hi @istvanszoboszlai and @AntonShipunov,

An example is provided within the `example` app of the webview_flutter plugin. The example loads the `index.html` page which internally has a relative reference to a style sheet called `styles.css`. 

- [Link](https://github.com/flutter/plugins/tree/main/packages/webview_flutter/webview_flutter/example) to the example project.
- [Link](https://github.com/flutter/plugins/tree/main/packages/webview_flutter/webview_flutter/example/assets/www) to the directory structure of the web files (inside the example project);
- [Configuration](https://github.com/flutter/plugins/blob/d07eaffb0d76ec128fa3e9746b1b624eda3483e2/packages/webview_flutter/webview_flutter/example/pubspec.yaml#L33) in the `pubspec.yaml` (this is a very important step, note that relative files/ paths need to be mentioned).
- [Example](https://github.com/flutter/plugins/blob/d07eaffb0d76ec128fa3e9746b1b624eda3483e2/packages/webview_flutter/webview_flutter/example/lib/main.dart#L404) usage in code.

**Comment by vinod-sutar-bbi** [comment]
the issue occured on my iPad which has iOS 12.5.1.

I have to make following changes in FlutterWebView.m in webview_flutter-1.0.7 dependency


```
- (bool)loadUrl:(NSString*)url withHeaders:(NSDictionary<NSString*, NSString*>*)headers {
  NSURL* nsUrl = [NSURL URLWithString:url];
  if (!nsUrl) {
    return false;
  }
  NSMutableURLRequest* request = [NSMutableURLRequest requestWithURL:nsUrl];
  [request setAllHTTPHeaderFields:headers];
    
    if([url hasPrefix:@"file"]) {
        
        NSString * allowingReadAccessToURL = [request.URL.path stringByDeletingLastPathComponent];
        
        if (@available(iOS 9.0, *)) {
            
            [_webView loadFileURL:request.URL allowingReadAccessToURL:[NSURL fileURLWithPath:allowingReadAccessToURL]];
        }
        else {
            
            [_webView loadRequest:request];
        }
    }
    else {
        
      [_webView loadRequest:request];
    }
    
//  [_webView loadRequest:request];
    
  return true;
}
```

**Comment by stuartmorgan** [comment]
I'm going to close this; the reported behavior that loading `file:` URLs doesn't allow relative loading is WKWebView behavior that we have no ability to change, so that approach cannot work, and as noted above the are now specific APIs for the use case of loading local and asset files where relative loads do work (as long as they are in the enclosing directory).
